hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/StateContext.java

```diff
@@ -112,7 +115,7 @@ public class StateContext {
   private final Map<InetSocketAddress, List<Message>>
       incrementalReportsQueue;
   private final Map<InetSocketAddress, Queue<ContainerAction>> containerActions;
-  private final Map<InetSocketAddress, Queue<PipelineAction>> pipelineActions;
+  private final Map<InetSocketAddress, LinkedHashMap<PipelineKey, PipelineAction>> pipelineActions;
```

Comment:
Let's also add a new class for the inner map.  It is easier to see the synchronization.
```java
  static class ActionMap {
    private final LinkedHashMap<PipelineKey, PipelineAction> map = new LinkedHashMap<>();

    synchronized int size() {
      return map.size();
    }

    synchronized void putIfAbsent(PipelineKey key, PipelineAction pipelineAction) {
      map.putIfAbsent(key, pipelineAction);
    }

    synchronized List<PipelineAction> getActions(List<PipelineReport> reports, int max) {
      if (map.isEmpty()) {
        return Collections.emptyList();
      }
      final List<PipelineAction> pipelineActionList = new ArrayList<>();
      final int limit = Math.min(map.size(), max);
      final Iterator<Map.Entry<PipelineKey, PipelineAction>> i = map.entrySet().iterator();
      for (int count = 0; count < limit && i.hasNext(); count++) {
        final Map.Entry<PipelineKey, PipelineAction> entry = i.next();
        final PipelineAction action = entry.getValue();

        // Add closePipeline back to the pipelineAction queue until
        // pipeline is closed and removed from the DN.
        if (action.hasClosePipeline()) {
          if (reports.stream().noneMatch(entry.getKey()::equalsId)) {
            // pipeline is removed from the DN, this action is no longer needed.
            i.remove();
            continue;
          }
          // pipeline is closed but not yet removed from the DN.
        } else {
          i.remove();
        }
        pipelineActionList.add(action);
      }
      // add all
      return pipelineActionList;
    }
  }
```